MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,247 @@
+  is within the last 70 minutes.
+* The loose object action doesn't trigger if the most-recent "fromloose" pack
+  timestamp is within the last 24 hours.
+* The prefetch repack action doesn't trigger if the most-recent "repacked" pack
+  timestamp is within the last 7 days or the most-recent "repacked" pack is
+  modified within the last 24 hours.
+
+In addition to the background maintenance, we can run the actions directly
+by creating a new `MaintenanceVerb` so commands like `gvfs maintenance [--loose|--packs]` will run the maintenance directly.
+
+### Watching Git Process IDs
+
+A possible direction for ensuring we don't delete a file that could be
+referenced by a Git process is to record a list of Git process IDs that are
+running as an action completes. We can then verify that those processes are
+not running the next time we run the action. See [`IsBlockingProcessRunning()`](https://github.com/Microsoft/VFSForGit/blob/4057d91a585b4f188874f8636b4390da402a367c/GVFS/GVFS.Common/InstallerPreRunChecker.cs#L129-L147)
+for an example of how to check the process list.
+
+We could centralize this Git process logic into an `ActionHelper` class that
+manages storing a "last-run information" file for each action, and that can
+include the previous-run timestamp and a list of Git processes. Storing the
+data in JSON could make this extensible in the future.
+
+Testing Strategy
+----------------
+
+We can test each action individually using functional tests.
+
+The `IsReady(time)` methods are simple to test, and allows us to check cases
+such as:
+
+* No previous time listed on the filesystem.
+* Time is old enough to return `true`.
+* Time is new enough to return `false`, both before and after `time`.
+
+The `Execute()` method requires a bit more work. We should focus on pre- and
+post-conditions to verify the actions do what we think they are doing, and
+satisfy the "contract" with Git that we don't delete files that Git may be
+reading at the time. We can also create tests that run some Git commands in
+parallel, but we can't control the timing of when the Git commands load the
+pack-files versus loose objects. The best option (that I can think of) is to
+run `git rev-list --objects --all` on loop until the `Execute()` method returns.
+This may catch unexpected behavior, but is likely to not fail consistently.
+Instead, the pre- and post-condition tests should be used to verify correctness
+more consistently.
+
+The `Stop()` method is important to test in the following states:
+
+* `Execute()` has not run.
+* `Execute()` has completed.
+* `Execute()` is running. This may require injecting "pauses" into the action
+  to check different stages during the execution. We could insert callbacks in
+  the action's constructor (that are `null` outside of tests) that allow us to
+  call `Stop()` at specific times in the `Execute()` method.
+
+How does the post-fetch job fit in?

```

Comment:
I had another idea here after I woke up this morning:

The post-fetch job could be the first action to leave a data file, and that file can store the last timestamp of a prefetch pack that was used in the commit-graph. We can then scan the pack-files to see which ones should be used. This can allow us to drop the message-passing algorithm, because we will do it in the background within 15 minutes. That changes the behavior slightly (we don't compute the multi-pack-index and commit-graph immediately after a `gvfs prefetch --commits`, but that is less and less likely to be a non-trivial action).